articles/avoid-fix-flaky-tests.md

```diff
@@ -0,0 +1,73 @@
+---
+authorName: Adrian Herzog
+authorGithubUsername: adiherzog
+issue: 42
+title: How to avoid / deal with flaky system tests
+---
+# {{page.title}}
+
+There are many reasons to avoid automated tests that check an entire system or even a group of connected systems. But there are also good reasons to build and maintain a reasonable number of such tests. For this article, let's assume that the tests are there for a good reason and provide valuable information to the team.
+
+Now, what can you do if one or more of these tests are flaky? By flaky I mean that the test changes from green to red or vice versa without any changes to the code.
+
+Here are a number of strategies that hopefully will help you keeping the tests stable.
+
+
+# Increase Your Chances To Succeed
+
+## Only Write Automated Tests For What's Under Your Control
+
+If the system under tests (SUT) containts components which are not under your control, it's very hard to keep the SUT in a stable state. For those systems not under your control, you might not know whether they have changed in the meantime or not. And even if you do know, you are most probably not able to influence them and change them back to how you need them.
+
+Therefore don't rely on other teams systems if possible. Mock them away so you have full control. Create System Tests instead of System Integration Tests. And of course these system tests must be written by the people who also program the system itself.
+
+If there's no way around System Integration Tests, try to get control over other systems as much as possible. E.g. align very closely with these systems regarding test data and deployments.
+
+## Avoid Broken Windows
+
+As soon as one test fails, insist on fixing it. Otherwise people might start ignoring failures and you'll end up in a [broken window](https://en.wikipedia.org/wiki/Broken_windows_theory) situation. There's a root cause to every flaky test. Either it's the test itself that needs improvement or it's a bug in the software.
+
+## Get The Timing Right
+
+Most issues with flaky tests are related to timing. Know your testing framework well, e.g. when it waits for something and when it doesn't. If you're using Selenium or Protractor, learn about [explicit and implicit waits](https://www.seleniumhq.org/docs/04_webdriver_advanced.jsp) and use them wisely (I strongly recommend using explicit waits). If your system under test has asynchronous behaviour, make sure you understand it well and write the tests accordingly. Configure sensible timeouts.
+
+
+# Analyze Failures Systematically
+
+## Reproduce Failures Locally
```

Comment:
This is a rather specialized form of the rules I follow when building CI systems.

Have your CI environment closely mirror your development environment. Ofcourse it all depends on the level of tests you're running: unit, integration, system etc.

I would invert the structure of the explanation:

The developer should be in a position to reproduce the failure (and the exact same execution steps) locally so that he/she can bring the full breadth of development tools to bear.
This requires that CI and test environment are structured accordingly.